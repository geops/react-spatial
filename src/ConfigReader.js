import TileLayer from 'ol/layer/Tile';
import XYZ from 'ol/source/XYZ';
import VectorSource from 'ol/source/Vector';
import WMTSSource from 'ol/source/WMTS';
import WMTSTileGrid from 'ol/tilegrid/WMTS';
import GeoJSONFormat from 'ol/format/GeoJSON';
import Layer from './Layer';
import VectorLayer from './VectorLayer';

import projections from './Projections';

class ConfigReader {
  static createXYZLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new XYZ({
          url: item.data.url,
        }),
      }),
    });
  }

  static createVectorLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new VectorLayer({
      ...conf,
      source: new VectorSource({
        url: item.data.url,
        format: new GeoJSONFormat(),
      }),
      style: item.data.style,
    });
  }

  static createWMTSLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
      olLayer: new TileLayer({
        zIndex: -1,
        source: new WMTSSource({
          url: item.data.url,
          tileGrid: new WMTSTileGrid({
            extent: item.data.projectionExtent,
            resolutions: item.data.resolutions,
            matrixIds: item.data.resolutions.map((res, i) => `${i}`),
          }),
        }),
      }),
    });
  }

  static createEmptyLayer(item) {
    const conf = { ...item };
    delete conf.data;

    return new Layer({
      ...conf,
    });
  }

  static createLayer(item) {
    let layer;

    switch (item.data.type) {
      case 'xyz':
        layer = ConfigReader.createXYZLayer(item);
        break;
      case 'wmts':
        layer = ConfigReader.createWMTSLayer(item);
        break;
      case 'vectorLayer':
        layer = ConfigReader.createVectorLayer(item);
        break;
      default:
        layer = ConfigReader.createEmptyLayer(item);
    }

    return layer;
  }

  constructor(map, data, style = null) {
    this.map = map;
    this.data = data;
    this.style = style;

    this.defaults = {
      data: {},
      visible: false,
      isBaseLayer: false,
    };
  }

  loadLayerFromConfig(config) {
    // apply default values
    const item = { ...this.defaults, ...config };

    if (item.data && item.data.type === 'wmts') {
      const proj = projections[item.data.projection];
      if (proj) {
        item.data.projectionExtent = proj.projectionExtent;
        item.data.resolutions = proj.resolutions;
      }
    }

    const layer = ConfigReader.createLayer(item);
    layer.init(this.map);

    if (item.children) {
      item.children.forEach(childConf => {
        layer.addChild(this.loadLayerFromConfig(childConf));
      });
    }

    return layer;
  }

  initialize() {
    return this.data.map(val => this.loadLayerFromConfig(val));
  }
}

const readConfig = (map, data) => {
  const configReader = new ConfigReader(map, data);
  return configReader.initialize();
};

const getVisibleTopic = topicList => {
  let visibleTopic = null;
  topicList.forEach(topic => {
    if (topic.visible) {
      visibleTopic = topic;
    }
  });
  return visibleTopic.children;
};

export default { getVisibleTopic, readConfig };
