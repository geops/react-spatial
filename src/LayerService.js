import Observable from 'ol/Observable';

/**
 * A layer service class to handle layer adding, removing and visiblity.
 */
export default class LayerService {
  constructor(layers) {
    this.layers = layers;
    this.callbacks = {};
    this.callbacks = {};
    this.keys = [];
    this.listenChangeEvt();
  }

  getLayers() {
    return this.layers;
  }

  getLayersAsFlatArray(optLayers) {
    let layers = [];
    (optLayers || this.getLayers() || []).forEach(l => {
      layers.push(l);
      const children = l.getChildren();
      layers = layers.concat(this.getLayersAsFlatArray(children));
    });
    return layers;
  }

  getLayer(name) {
    return this.getLayersAsFlatArray().find(l => l.getName() === name);
  }

  getParent(child) {
    return this.getLayersAsFlatArray().find(
      l => !!l.getChildren().includes(child),
    );
  }

  getRadioGroupLayers(radioGroupName) {
    if (radioGroupName) {
      return this.getLayersAsFlatArray().filter(
        l => l.getRadioGroup() === radioGroupName,
      );
    }

    return null;
  }

  on(evt, callback) {
    this.callbacks[evt] = this.callbacks[evt] || [];
    this.callbacks[evt].push(callback);
  }

  listenChangeEvt() {
    this.getLayersAsFlatArray().forEach(layer => {
      this.keys.push(
        layer.on('change:visible', evt => {
          const visible = evt.target.getVisible();

          // Apply to siblings only if it's a radio group.
          if (
            !evt.stopPropagationSiblings &&
            layer.getRadioGroup() &&
            visible
          ) {
            const siblings = this.getRadioGroupLayers(
              layer.getRadioGroup(),
            ).filter(l => l !== layer);

            siblings.forEach(s => {
              if (
                visible &&
                s.getRadioGroup() &&
                evt.target.getRadioGroup() === s.getRadioGroup()
              ) {
                s.setVisible(false, false, true, true);
              }
            });
          }

          // Apply to children
          if (!evt.stopPropagationDown && layer.children) {
            layer.children.forEach(child => {
              child.setVisible(visible, false, true, false);
            });
          }

          // Apply to parent only if:
          //   - a child is visible
          //   - all children are hidden
          const parentLayer = this.getParent(layer);

          if (
            !evt.stopPropagationUp &&
            parentLayer &&
            (visible ||
              (!visible && !parentLayer.children.find(c => c.getVisible())))
          ) {
            parentLayer.setVisible(visible, true, false, false);
          }

          (this.callbacks['change:visible'] || []).forEach(cb => cb());
        }),
      );
    });
  }

  unlistenChangeEvt() {
    this.keys.forEach(key => {
      Observable.unByKey(key);
    });
  }
}
