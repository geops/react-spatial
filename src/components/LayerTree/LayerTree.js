import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Checkbox from '../Checkbox';
import Button from '../Button';

const propTypes = {
  /**
   * Layers provider.
   */
  layerService: PropTypes.object,

  /**
   * CSS class to apply on the container.
   */
  className: PropTypes.string,

  /**
   * CSS class to apply on each item.
   */
  classNameItem: PropTypes.string,

  /**
   * CSS class to apply to the label element which contains the input.
   * Unused if you use `renderItem` property.
   */
  classNameInput: PropTypes.string,

  /**
   * CSS class to apply to the toggle button which contains the title and the arrow.
   * Unused if you use `renderItem` property.
   */
  classNameToggle: PropTypes.string,

  /**
   * CSS class to apply to the arrow.
   * Unused if you use `renderItem` property.
   */
  classNameArrow: PropTypes.string,

  /**
   * Padding left to apply on each level.
   */
  padding: PropTypes.number,

  /**
   * Determine if the item is hidden in the tree or not.
   *
   * @param {object} item The item to hide or not.
   *
   * @return {bool} true if the item is not displayed in the tree
   */
  isItemHidden: PropTypes.func,

  /**
   * Custom function to render an item in the tree.
   *
   * @param {object} item The item to render.
   *
   * @return {node} A jsx node.
   */
  renderItem: PropTypes.func,
};

const defaultProps = {
  layerService: undefined,
  className: 'tm-layer-tree',
  classNameItem: 'tm-layer-tree-item',
  classNameInput: 'tm-layer-tree-input',
  classNameToggle: 'tm-layer-tree-toggle',
  classNameArrow: 'tm-layer-tree-arrow',
  padding: 30,
  isItemHidden: () => false,
  renderItem: null,
};

class LayerTree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      layers: props.layerService ? props.layerService.getLayers() : [],
      expandedLayerNames: [],
    };

    this.olKeys = [];
  }

  componentDidMount() {
    this.updateLayerService();
  }

  componentDidUpdate(prevProps) {
    const { layerService } = this.props;

    if (layerService !== prevProps.layerService) {
      this.updateLayerService();
    }
  }

  onInputClick(layer, toggle = false) {
    if (toggle) {
      this.onToggle(layer);
    } else {
      layer.setVisible(!layer.getVisible());
    }
  }

  onToggle(layer) {
    const { expandedLayerNames } = this.state;
    const pos = expandedLayerNames.indexOf(layer.getName());
    if (pos > -1) {
      expandedLayerNames.splice(pos, 1);
    } else {
      expandedLayerNames.push(layer.getName());
    }

    this.setState({ expandedLayerNames });
  }

  updateLayerService() {
    const { layerService } = this.props;
    if (layerService) {
      this.updateLayers(layerService);
      layerService.on('change:visible', () => this.updateLayers(layerService));
    }
  }

  updateLayers(layerService) {
    this.setState({
      layers: layerService.getLayers(),
    });
  }

  renderInput(layer) {
    const { classNameInput } = this.props;
    let tabIndex = 0;

    if (!layer.getChildren().length) {
      // We forbid focus on keypress event for first level layers and layers without children.
      tabIndex = -1;
    }

    const inputType = layer.getRadioGroup() ? 'radio' : 'checkbox';
    return (
      <Checkbox
        tabIndex={tabIndex}
        inputType={inputType}
        checked={layer.getVisible()}
        className={`${classNameInput} ${classNameInput}-${inputType}`}
        onClick={() => this.onInputClick(layer)}
      />
    );
  }

  renderArrow(layer) {
    const { classNameArrow } = this.props;
    const { expandedLayerNames } = this.state;

    if (!layer.getChildren().length) {
      return null;
    }

    return (
      <div
        className={`${classNameArrow} ${classNameArrow}${
          expandedLayerNames.includes(layer.getName())
            ? '-collapsed'
            : '-expanded'
        }`}
      />
    );
  }

  // Render a button which expands/collapse the layer if there is children
  // or simulate a click on the input otherwise.
  renderToggleButton(layer) {
    const { classNameToggle } = this.props;
    const tabIndex = 0;

    return (
      <Button
        tabIndex={tabIndex}
        className={classNameToggle}
        onClick={() => {
          this.onInputClick(layer, layer.getChildren().length);
        }}
      >
        <div>{layer.getName()}</div>
        {this.renderArrow(layer)}
      </Button>
    );
  }

  renderItem(layer, level) {
    const { renderItem, classNameItem, padding } = this.props;
    const { expandedLayerNames } = this.state;

    const children = expandedLayerNames.includes(layer.getName())
      ? []
      : [...layer.getChildren()];

    if (renderItem) {
      return renderItem(layer);
    }

    return (
      <div key={layer.getName()}>
        <div
          className={classNameItem}
          style={{
            paddingLeft: `${padding * level}px`,
          }}
        >
          {this.renderInput(layer)}
          {this.renderToggleButton(layer)}
        </div>
        {[...children]
          .reverse()
          .map(child => this.renderItem(child, level + 1))}
      </div>
    );
  }

  renderTree() {
    const { isItemHidden } = this.props;
    const { layers } = this.state;

    if (!layers) {
      return null;
    }

    return (
      <>
        {layers
          .filter(l => !isItemHidden(l))
          .reverse()
          .map(l => this.renderItem(l, 0))}
      </>
    );
  }

  render() {
    const { className } = this.props;
    return <div className={className}>{this.renderTree()}</div>;
  }
}

LayerTree.propTypes = propTypes;
LayerTree.defaultProps = defaultProps;

export default LayerTree;
