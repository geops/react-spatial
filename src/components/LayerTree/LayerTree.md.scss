.tm-layer-tree-example {
  position: relative;

  .tm-map {
    width: 100%;
    height: 100%;
    flex-grow: 1;
  }

  .tm-layer-tree {
    background: white;
    padding: 10px 5px;
    width: 250px;
    top: 10px;
    left: 10px;
    bottom: 10px;
    position: absolute;
    margin: auto;
  }
}
