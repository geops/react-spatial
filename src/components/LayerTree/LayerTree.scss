.tm-layer-tree {
  width: 100%;
  overflow-y: auto;

  .tm-layer-tree-arrow {
    border: solid black;
    border-width: 0 1px 1px 0;
    display: inline-block;
    padding: 0;
    background: transparent;
    width: 5px;
    height: 5px;
  }

  .tm-layer-tree-arrow-collapsed {
    transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
    margin-top: -3px;
  }

  .tm-layer-tree-arrow-expanded {
    transform: rotate(-135deg);
    -webkit-transform: rotate(-135deg);
    margin-top: 3px;
  }

  .tm-layer-tree-item {
    display: flex;
    align-items: center;
    padding: 10px;
    position: relative;

    & > * {
      margin-right: 10px;
      align-items: center;
    }

    div {
      /* HACK: Without this it breaks the click on the label (Chrome) */
      position: relative;
    }

    /* CSS for the toggle div */
    .tm-layer-tree-toggle {
      width: auto;
      height: auto;
      flex-grow: 1;
      display: flex;
      align-items: center;
      justify-content: left;

      div:first-child {
        margin-right: 10px;
      }
    }

    /* Customize the label (the container) */
    label {
      display: inline-block;
      position: relative;
      cursor: pointer;
    }

    label,
    label input {
      height: 18px;
      width: 18px;
    }

    label input,
    label span {
      position: absolute;
      top: 0;
      left: 0;
      margin: 0;
    }

    label input {
      display: none;
    }

    /* Create a custom checkbox */
    label span {
      background-color: white;
      border: 1px solid lightgray;
      width: calc(100% - 2px);
      height: calc(100% - 2px);
    }

    .tm-layer-tree-input-radio span {
      border-radius: 50%;
    }

    /* On mouse-over, add a grey background color */
    label:hover input ~ span {
      background-color: #ccc;
    }

    /* Create the checkmark/indicator (hidden when not checked) */
    label span::after {
      content: '';
      position: absolute;
      display: none;
    }

    /* Style the checkmark/indicator */
    .tm-layer-tree-input-checkbox span::after {
      left: 5px;
      top: 1px;
      width: 5px;
      height: 10px;
      border: solid $brand-primary;
      border-width: 0 1px 1px 0;
      -webkit-transform: rotate(45deg);
      -ms-transform: rotate(45deg);
      transform: rotate(45deg);
    }

    /* Style the  (dot/circle) */
    .tm-layer-tree-input-radio span::after {
      top: 2px;
      left: 2px;
      width: 12px;
      height: 12px;
      border-radius: 50%;
      background: $brand-primary;
    }

    label input:checked ~ span::after {
      display: block;
    }
  }
}
