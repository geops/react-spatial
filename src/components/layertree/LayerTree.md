#

This demonstrates the use of LayerTree.

```jsx
const React = require('react');
const LayerTree = require('./LayerTree').default;
const BasicMap = require('../map/BasicMap').default;
const VectorLayer = require('../../VectorLayer').default;
const LayerService = require('../../LayerService').default;
const OLMap = require('ol/Map').default;
const Feature = require('ol/Feature').default;
const Point = require('ol/geom/Point').default;
const VectorSource = require('ol/source/Vector').default;
const Style = require('ol/style/Style').default;
const Circle = require('ol/style/Circle').default;
const Fill = require('ol/style/Fill').default;
const Text = require('ol/style/Text').default;
const ConfigReader = require('../../ConfigReader').default;
require('./LayerTree.md.scss');

class LayerTreeExample extends React.Component {
  constructor(props) {
    super(props);

    this.center = [-10997148, 4569099];
    this.map = new OLMap({controls:[]});
    this.state = {
      layerService: null
    }
  }

  componentDidMount() {
    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );
    this.setState({
      layerService: new LayerService(layers)
    })
  }

  applyStyle(itemId) {
    this.layers[0].olLayer.getStyle().setText(
      new Text({
        font: '20px sans-serif',
        text: 'Last item modified is : ' + itemId,
      }),
    );
    this.layers[0].olLayer.changed();
  }

  render() {
    const {layerService} = this.state;
    return (
      <div className="tm-layer-tree-example">
        <LayerTree
          layerService={layerService}
        />
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={3}
          layers={this.layers}
        />
      </div>
    );
  }
}

<LayerTreeExample />;
```
