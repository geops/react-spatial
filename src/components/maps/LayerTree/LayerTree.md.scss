.rs-layer-tree-example {
  position: relative;
  height: 300px;

  .rs-map {
    width: 100%;
    height: 100%;
    flex-grow: 1;
  }

  .rs-layer-tree {
    background: white;
    padding: 10px 5px;
    width: 250px;
    top: 10px;
    left: 10px;
    bottom: 10px;
    position: absolute;
    margin: auto;
  }
}
