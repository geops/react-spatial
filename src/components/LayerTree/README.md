#

This demonstrates the use of LayerTree.

```jsx
import React from  'react'
import LayerTree from 'react-spatial/components/LayerTree';
import BasicMap from 'react-spatial/components/BasicMap';
import LayerService from 'react-spatial/LayerService';
import OLMap from 'ol/Map';
import ConfigReader from 'react-spatial/ConfigReader';

class LayerTreeExample extends React.Component {
  constructor(props) {
    super(props);

    this.center = [-10997148, 4569099];
    this.map = new OLMap({controls:[]});
    this.state = {
      layerService: null
    }
  }

  componentDidMount() {
    const layers = ConfigReader.readConfig(
      this.map,
      treeData,
    );
    this.setState({
      layerService: new LayerService(layers)
    })
  }

  render() {
    const {layerService} = this.state;
    return (
      <div className="tm-layer-tree-example">
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={3}
          layers={this.layers}
        />
        <LayerTree
          layerService={layerService}
        />
      </div>
    );
  }
}

<LayerTreeExample />;
```
