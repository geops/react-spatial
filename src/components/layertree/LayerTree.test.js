import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-canvas-mock';
import OLMap from 'ol/Map';
import renderer from 'react-test-renderer';
import LayerTree from './LayerTree';
import data from '../../../data/TreeData';
import ConfigReader from '../../ConfigReader';
import LayerService from '../../LayerService';

configure({ adapter: new Adapter() });

const mountLayerTree = newData => {
  const layers = ConfigReader.readConfig(new OLMap({}), newData);
  const layerService = new LayerService(layers);
  return mount(<LayerTree layerService={layerService} />);
};

const renderLayerTree = (newData, props) => {
  const layers = ConfigReader.readConfig(new OLMap({}), newData);
  const layerService = new LayerService(layers);
  const component = renderer.create(
    <LayerTree layerService={layerService} {...props || {}} />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
};

const classItem = '.tm-layer-tree-item';
const toggleItem = '.tm-layer-tree-toggle';

describe('LayerTree', () => {
  describe('matches snapshots', () => {
    test('using default properties.', () => {
      renderLayerTree(data);
    });

    test('when no layers.', () => {
      const layerService = new LayerService();
      const component = renderer.create(
        <LayerTree layerService={layerService} />,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('when renderItem is used.', () => {
      renderLayerTree(data, {
        renderItem: item => <div key={item.getName()}>{item.getName()}</div>,
      });
    });

    test('when classNames are used.', () => {
      renderLayerTree(data, {
        className: 'foo',
        classNameItem: 'bar',
        classNameInput: 'qux',
        classNameToggle: 'quux',
        classNameArrow: 'ged',
      });
    });

    test('when an item is hidden.', () => {
      renderLayerTree(data, {
        isItemHidden: item => !!item.children.length,
      });
    });
  });

  describe('triggers onInputClick', () => {
    let wrapper;
    let spy;
    let spy2;
    const data2 = [
      {
        name: 'foo',
        data: {
          type: 'xyz',
        },
      },
    ];
    const expectCalled = () => {
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy2).toHaveBeenCalledTimes(0);
      expect(spy.mock.calls[0][0].getName()).toBe('foo');
    };

    beforeEach(() => {
      spy = jest.spyOn(LayerTree.prototype, 'onInputClick');
      spy2 = jest.spyOn(LayerTree.prototype, 'onToggle');
      wrapper = mountLayerTree(data2);
    });

    afterEach(() => {
      spy.mockRestore();
      spy2.mockRestore();
    });

    test('when we press enter with keyboard on the label element.', () => {
      wrapper
        .find('label')
        .at(0)
        .simulate('keypress', { which: 13 });
      expectCalled();
    });

    test('when we click on input.', () => {
      wrapper
        .find('input')
        .at(0)
        .simulate('click');
      expectCalled();
    });

    test('when we click on toggle button (label+arrow) of an item without children.', () => {
      wrapper
        .find(classItem)
        .first()
        .childAt(1)
        .simulate('click');
      expectCalled();
    });
  });

  describe('triggers onToggle', () => {
    let wrapper;
    let spy;
    const data2 = [
      {
        name: '1',
        children: [
          {
            name: '1-1',
          },
          {
            name: '1-1-1',
            data: {
              type: 'xyz',
            },
          },
        ],
      },
    ];

    const expectCalled = () => {
      expect(spy).toHaveBeenCalledTimes(1);
    };

    beforeEach(() => {
      spy = jest.spyOn(LayerTree.prototype, 'onToggle');
      wrapper = mountLayerTree(data2);
    });

    test('when we click on toggle button (label+arrow) of an item with children', () => {
      wrapper
        .find(toggleItem)
        .first()
        .simulate('click');
      expectCalled();
    });
  });
});
