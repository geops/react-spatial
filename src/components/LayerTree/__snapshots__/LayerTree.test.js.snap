// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerTree matches snapshots using default properties. 1`] = `
<div
  className="tm-layer-tree"
>
  <div>
    <div
      className="tm-layer-tree-item"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="tm-layer-tree-input tm-layer-tree-input-radio tm-radio"
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="radio"
        />
        <span />
      </label>
      <div
        className="tm-layer-tree-toggle"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          Vector sample layers
        </div>
        <div
          className="tm-layer-tree-arrow tm-layer-tree-arrow-expanded"
        />
      </div>
    </div>
    <div>
      <div
        className="tm-layer-tree-item"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="tm-layer-tree-input tm-layer-tree-input-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="tm-layer-tree-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Points Samples
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="tm-layer-tree-item"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="tm-layer-tree-input tm-layer-tree-input-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={true}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="tm-layer-tree-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Lines Samples
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="tm-layer-tree-item"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="tm-layer-tree-input tm-layer-tree-input-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="tm-layer-tree-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Polygons Samples
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="tm-layer-tree-item"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="tm-layer-tree-input tm-layer-tree-input-checkbox tm-checkbox"
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <span />
      </label>
      <div
        className="tm-layer-tree-toggle"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          Others layers
        </div>
        <div
          className="tm-layer-tree-arrow tm-layer-tree-arrow-expanded"
        />
      </div>
    </div>
    <div>
      <div
        className="tm-layer-tree-item"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="tm-layer-tree-input tm-layer-tree-input-checkbox tm-checkbox"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <span />
        </label>
        <div
          className="tm-layer-tree-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Countries Borders
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="tm-layer-tree-item"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="tm-layer-tree-input tm-layer-tree-input-checkbox tm-checkbox"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={true}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <span />
        </label>
        <div
          className="tm-layer-tree-toggle"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            USA Population Density
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="tm-layer-tree-item"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="tm-layer-tree-input tm-layer-tree-input-checkbox tm-checkbox"
        onKeyPress={[Function]}
        tabIndex={-1}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <span />
      </label>
      <div
        className="tm-layer-tree-toggle"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          OSM Baselayer
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayerTree matches snapshots when an item is hidden. 1`] = `
<div
  className="tm-layer-tree"
>
  <div>
    <div
      className="tm-layer-tree-item"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="tm-layer-tree-input tm-layer-tree-input-checkbox tm-checkbox"
        onKeyPress={[Function]}
        tabIndex={-1}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <span />
      </label>
      <div
        className="tm-layer-tree-toggle"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          OSM Baselayer
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayerTree matches snapshots when classNames are used. 1`] = `
<div
  className="foo"
>
  <div>
    <div
      className="bar"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="qux qux-radio tm-radio"
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="radio"
        />
        <span />
      </label>
      <div
        className="quux"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          Vector sample layers
        </div>
        <div
          className="ged ged-expanded"
        />
      </div>
    </div>
    <div>
      <div
        className="bar"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="qux qux-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="quux"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Points Samples
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="bar"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="qux qux-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={true}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="quux"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Lines Samples
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="bar"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="qux qux-radio tm-radio"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="radio"
          />
          <span />
        </label>
        <div
          className="quux"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Polygons Samples
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="bar"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="qux qux-checkbox tm-checkbox"
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <span />
      </label>
      <div
        className="quux"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          Others layers
        </div>
        <div
          className="ged ged-expanded"
        />
      </div>
    </div>
    <div>
      <div
        className="bar"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="qux qux-checkbox tm-checkbox"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={false}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <span />
        </label>
        <div
          className="quux"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            Countries Borders
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        className="bar"
        style={
          Object {
            "paddingLeft": "30px",
          }
        }
      >
        <label
          className="qux qux-checkbox tm-checkbox"
          onKeyPress={[Function]}
          tabIndex={-1}
        >
          <input
            checked={true}
            onChange={[Function]}
            onClick={[Function]}
            tabIndex={-1}
            type="checkbox"
          />
          <span />
        </label>
        <div
          className="quux"
          onClick={[Function]}
          onKeyPress={[Function]}
          role="button"
          tabIndex={0}
        >
          <div>
            USA Population Density
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="bar"
      style={
        Object {
          "paddingLeft": "0px",
        }
      }
    >
      <label
        className="qux qux-checkbox tm-checkbox"
        onKeyPress={[Function]}
        tabIndex={-1}
      >
        <input
          checked={true}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex={-1}
          type="checkbox"
        />
        <span />
      </label>
      <div
        className="quux"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div>
          OSM Baselayer
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LayerTree matches snapshots when no layers. 1`] = `
<div
  className="tm-layer-tree"
/>
`;

exports[`LayerTree matches snapshots when renderItem is used. 1`] = `
<div
  className="tm-layer-tree"
>
  <div>
    Vector sample layers
  </div>
  <div>
    Others layers
  </div>
  <div>
    OSM Baselayer
  </div>
</div>
`;
